Move bwcVersions extension property to BuildParams

breskeby opened the pull request:
- resolve bwcVersions in GlobalBuildInfoPlugin
- related / prerequisite for https://github.com/elastic/elasticsearch/pull/56003
- TODO: probably more bwc related work can be moved from build scripts into
certain plugins.

breskeby committed:
Move bwcVersions extension property to BuildParams

- resolved in GlobalBuildInfoPlugin
- TODO: probably more bwc related work can be moved from build scripts into
certain plugins.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Build)

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:
```diff
@@ -92,12 +97,26 @@ public void apply(Project project) {
+            List<String> versionLines = IOUtils.readLines(new FileInputStream(versionsFile), "UTF-8");
+            return new BwcVersions(versionLines);
+        } catch (IOException e) {
+            return null;
```
I think we should rethrow as an `UncheckedIOException`. If we aren't able to determine BWC versions this is likely to cause all sorts of build issues so we should treat this as a fatal error. I believe this might already be the case as `BuildParams` will validate we aren't setting the value to `null`, but then we end up losing the original exception cause which is not ideal.

breskeby replied:
I thought about it but didn’t want to fail fatal here to make testing easier. I needed to touch the testkit Test builds and wanted to keep the ability to direct configure the bwcVersion in via BuildParamd there instead of introducing sth like a test only override property in the plugin itself to pass a configurable file path to resolve versions from

breskeby replied:
I thought about it but didn’t want to fail fatal here to make testing easier. I needed to touch the testkit Test builds and wanted to keep the ability to direct configure the bwcVersion in via BuildParamd there instead of introducing sth like a test only override property in the plugin itself to pass a configurable file path to resolve versions from

mark-vieira replied:
Gotcha, we should at least log the error then, so if it does barf we get a better message than just `bwcVersions cannot be null` or whatever.

rjernst replied:
Why does not failing with an exception make this easier to test? Tests can use expectThrows to catch an exception?

mark-vieira replied:
That doesn't work with testKit tests since it's not the _test_ that's throwing the exception, it's the build and those don't get propagated to the test. You just get a "your build failed" type exception returned by teskit.

mark-vieira reviewed (approved):
One comment on error handling but otherwise LGTM.

breskeby commented:
I thought about it but didn’t want to fail fatal here to make testing easier. I needed to touch the testkit Test builds and wanted to keep the ability to direct configure the bwcVersion in via BuildParamd there instead of introducing sth like a test only override property in the plugin itself to pass a configurable file path to resolve versions from

> On 5. May 2020, at 18:50, Mark Vieira <notifications@github.com> wrote:
> 
> ﻿
> @mark-vieira approved this pull request.
> 
> One comment on error handling but otherwise LGTM.
> 
> In buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:
> 
> >          });
>  
>          // Print global build info header just before task execution
>          project.getGradle().getTaskGraph().whenReady(graph -> logGlobalBuildInfo());
>      }
>  
> +    /* Introspect all versions of ES that may be tested against for backwards
> +     * compatibility. It is *super* important that this logic is the same as the
> +     * logic in VersionUtils.java. */
> +    private static BwcVersions resolveBwcVersions(File root) {
> +        File versionsFile = new File(root, DEFAULT_VERSION_JAVA_FILE_PATH);
> +        try {
> +            List<String> versionLines = IOUtils.readLines(new FileInputStream(versionsFile), "UTF-8");
> +            return new BwcVersions(versionLines);
> +        } catch (IOException e) {
> +            return null;
> I think we should rethrow as an UncheckedIOException. If we aren't able to determine BWC versions this is likely to cause all sorts of build issues so we should treat this as a fatal error. I believe this might already be the case as BuildParams will validate we aren't setting the value to null, but then we end up losing the original exception cause which is not ideal.
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

breskeby commented:
The problem is not about testing the exception handling but tweak testing scenarios when e.g. testing the DistributionDownloadPlugin by injecting custom versions

> On 6. May 2020, at 00:30, Ryan Ernst <notifications@github.com> wrote:
> 
> ﻿
> @rjernst commented on this pull request.
> 
> In buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:
> 
> >          });
>  
>          // Print global build info header just before task execution
>          project.getGradle().getTaskGraph().whenReady(graph -> logGlobalBuildInfo());
>      }
>  
> +    /* Introspect all versions of ES that may be tested against for backwards
> +     * compatibility. It is *super* important that this logic is the same as the
> +     * logic in VersionUtils.java. */
> +    private static BwcVersions resolveBwcVersions(File root) {
> +        File versionsFile = new File(root, DEFAULT_VERSION_JAVA_FILE_PATH);
> +        try {
> +            List<String> versionLines = IOUtils.readLines(new FileInputStream(versionsFile), "UTF-8");
> +            return new BwcVersions(versionLines);
> +        } catch (IOException e) {
> +            return null;
> Why does not failing with an exception make this easier to test? Tests can use expectThrows to catch an exception?
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

mark-vieira commented:
Yeah, since this is now done by a plugin instead of the by root build script those plugin tests include this logic so the only other way to override would be to create a `Version.java` file in the appropriate place or include some other way to override this behavior for tests.

That brings up a good point though. Will external users of `GlobalBuildInfoPlugin` barf if we can't load bwc versions? I think it will since things like the `PluginBuildPlugin` end up applying the `GlobalBuildInfoPlugin`.

I think we have a couple options here:

1. In the case where bwc information is unavailable, fail on _read_ instead of eagerly when loading the info.
2. Move this into a separate plugin, and apply that plugin only to projects which require bwc version information.

rjernst commented:
> Will external users of GlobalBuildInfoPlugin barf if we can't load bwc versions?

We have a guard on this already. In all the places we use bwcVersions, we check isInternal. This is something we should fix, either by having a buildSrc build time copy of Version.java into buildSrc resources, or read Version.java at that time and generating a json resource file that can be loaded more directly, but that we don't need to worry about bad formatting or anything missing (my preferred approach).

mark-vieira commented:
> This is something we should fix, either by having a buildSrc build time copy of Version.java into buildSrc resources, or read Version.java at that time and generating a json resource file that can be loaded more directly

What would be the benefit of this if we never intend external users to need BWC information?

rjernst commented:
The bwc information is needed to run any kind of bwc tests. It's reasonable to assume external plugin authors will want to test compatibility of their own plugin against other versions when we finally break lockstep plugin versioning.

mark-vieira commented:
Sure, but right now we have no kind of guarantees on how this works, and there exists no documentation or guidance on how one might integration BWC testing into their plugin build.

I'd say until we officially support such a thing we make bwcversions a purely internal mechanism (guarded like it is as you mention) and simply blow up if you try to access it outside the Elasticsearch build.

rjernst commented:
That's fair, but don't external plugin authors still load GlobalBuildInfoPlugin? How can we blow up in trying to load it outside of ES without causing all plugin authors builds to fail?

mark-vieira commented:
We do as Rene has done here and gracefully handle a missing `Version.java` file on load, and instead assertNotNull on read.

rjernst commented:
Can't we use `isInternal` to determine whether we should fail?

mark-vieira commented:
> Can't we use isInternal to determine whether we should fail?

Good call. @rene, let's do what @rjernst suggests. If we detect that this is an "internal" build (meaning we are running in the elasticsearch project vs the `build-tools` jar being pulled in by an external project) and we fail to parse `Version.java` for any reason let's explode, otherwise return `null` as we do now and then fail on read.

breskeby commented:
Will do that check on isInternal. Good idea and thanks for the discussion here. 


> On 6. May 2020, at 04:33, Mark Vieira <notifications@github.com> wrote:
> 
> ﻿
> Can't we use isInternal to determine whether we should fail?
> 
> Good call. @rene, let's do what @rjernst suggests. If we detect that this is an "internal" build (meaning we are running in the elasticsearch project vs the build-tools jar being pulled in by an external project) and we fail to parse Version.java for any reason let's explode, otherwise return null as we do now and then fail on read.
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

breskeby committed:
Resolve bwcVersions only for internal usage of GlobalBuildInfoPlugin

- propagate exception if version file cannot be resolved
- add null check when setting bwcVersions

breskeby commented:
I tweaked the implementation to only set the BuildParams.bwcVersions property for internal builds and _fail_ early when 1. sth goes wrong in parsing the version file and 2. if a passed bwcVersion to BuildParams.setBwcVersion(...) is null. IMO this is the most robust approach.

breskeby commented:
@elasticmachine update branch

breskeby commented:
@elasticmachine test this please

breskeby commented:
@elasticmachine update branch

breskeby commented:
@elasticmachine test this please

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:
```diff
@@ -69,10 +73,13 @@ public void apply(Project project) {
-        // Initialize global build parameters
         BuildParams.init(params -> {
+            // Initialize global build parameters
+            boolean internalUse = GlobalBuildInfoPlugin.class.getResource("/buildSrc.marker") != null;
```
nit: I would name this isInternal since that is the property we are setting

rjernst reviewed (approved):
LGTM

mark-vieira commented:
Changes LGTM.

breskeby committed:
Apply small review feedback on variable name

breskeby committed:
Fix BwcVersions java8 compatibility after merge

mark-vieira commented:
@breskeby can you please ensure this change is backported to `7.x`, `7.8` and `7.7` branches, and appropriate version labels are added to this PR?

breskeby commented:
7.x Backport PR: https://github.com/elastic/elasticsearch/pull/56381

breskeby commented:
7.7 backport PR https://github.com/elastic/elasticsearch/pull/56382

breskeby commented:
7.7 backport PR https://github.com/elastic/elasticsearch/pull/56383
